LLVM 23 breaks tests/codegen-llvm/char-ascii-branchless.rs

After https://github.com/llvm/llvm-project/pull/183692 landed upstream, we started seeing tests/codegen-llvm/char-ascii-branchless.rs fail:

```
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/tests/codegen-llvm/char-ascii-branchless.rs:38:16: error: CHECK-NOT: excluded string found in input
--
// CHECK-NOT: br {{.*}}
^
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/test/codegen-llvm/char-ascii-branchless/char-ascii-branchless.ll:64:2: note: found here
br i1 %_8.i, label %bb6.i, label %_RNvMNtNtCscGzNLIcHm04_4core4char7methodsc20is_ascii_punctuation.exit
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/tests/codegen-llvm/char-ascii-branchless.rs:45:16: error: CHECK-NOT: excluded string found in input
// CHECK-NOT: br {{.*}}
^
/var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/test/codegen-llvm/char-ascii-branchless/char-ascii-branchless.ll:88:2: note: found here
br i1 %_8.i, label %bb6.i, label %_RNvMs4_NtCscGzNLIcHm04_4core3numh20is_ascii_punctuation.exit
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
Input file: /var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/build/x86_64-unknown-linux-gnu/test/codegen-llvm/char-ascii-branchless/char-ascii-branchless.ll
Check file: /var/lib/buildkite-agent/builds/rust-llvm-integrate/llvm-project/rust-llvm-integrate-prototype/tests/codegen-llvm/char-ascii-branchless.rs
 
-dump-input=help explains the following input dump.
 
Input was:
<<<<<<
1: ; ModuleID = 'char_ascii_branchless.115aba9230e3e679-cgu.0'
2: source_filename = "char_ascii_branchless.115aba9230e3e679-cgu.0"
3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
4: target triple = "x86_64-unknown-linux-gnu"
5:
6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
7: define noundef zeroext i1 @is_ascii_alphanumeric_char(i32 noundef range(i32 0, 1114112) %x) unnamed_addr #0 {
8: start:
9:  %0 = add nsw i32 %x, -48
10:  %_4.sroa.0.0 = icmp ult i32 %0, 10
11:  %_5 = icmp samesign ugt i32 %x, 64
12:  %_7 = icmp samesign ugt i32 %x, 96
13:  %spec.select.v = select i1 %_7, i32 123, i32 91
14:  %spec.select = icmp samesign ult i32 %x, %spec.select.v
15:  %_8.sroa.0.0 = select i1 %_5, i1 %spec.select, i1 %_4.sroa.0.0
16:  ret i1 %_8.sroa.0.0
17: }
18:
19: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
20: define noundef zeroext i1 @is_ascii_alphanumeric_u8(i8 noundef %x) unnamed_addr #0 {
21: start:
22:  %0 = add nsw i8 %x, -48
23:  %_4.sroa.0.0 = icmp ult i8 %0, 10
24:  %_5 = icmp ugt i8 %x, 64
25:  %_7 = icmp ugt i8 %x, 96
26:  %spec.select.v = select i1 %_7, i8 123, i8 91
27:  %spec.select = icmp ult i8 %x, %spec.select.v
28:  %_8.sroa.0.0 = select i1 %_5, i1 %spec.select, i1 %_4.sroa.0.0
29:  ret i1 %_8.sroa.0.0
30: }
31:
32: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
33: define noundef zeroext i1 @is_ascii_hexdigit_char(i32 noundef range(i32 0, 1114112) %x) unnamed_addr #0 {
34: start:
35:  %0 = add nsw i32 %x, -48
36:  %_4.sroa.0.0 = icmp ult i32 %0, 10
37:  %_5 = icmp samesign ugt i32 %x, 64
38:  %_7 = icmp samesign ugt i32 %x, 96
39:  %spec.select.v = select i1 %_7, i32 103, i32 71
40:  %spec.select = icmp samesign ult i32 %x, %spec.select.v
41:  %_8.sroa.0.0 = select i1 %_5, i1 %spec.select, i1 %_4.sroa.0.0
42:  ret i1 %_8.sroa.0.0
43: }
44:
45: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
46: define noundef zeroext i1 @is_ascii_hexdigit_u8(i8 noundef %x) unnamed_addr #0 {
47: start:
48:  %0 = add nsw i8 %x, -48
49:  %_4.sroa.0.0 = icmp ult i8 %0, 10
50:  %_5 = icmp ugt i8 %x, 64
51:  %_7 = icmp ugt i8 %x, 96
52:  %spec.select.v = select i1 %_7, i8 103, i8 71
53:  %spec.select = icmp ult i8 %x, %spec.select.v
54:  %_8.sroa.0.0 = select i1 %_5, i1 %spec.select, i1 %_4.sroa.0.0
55:  ret i1 %_8.sroa.0.0
56: }
57:
58: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
59: define noundef zeroext i1 @is_ascii_punctuation_char(i32 noundef range(i32 0, 1114112) %0) unnamed_addr #0 {
60: start:
61:  %1 = add nsw i32 %0, -33
62:  %_4.sroa.0.0.i = icmp ult i32 %1, 15
63:  %_8.i = icmp samesign ugt i32 %0, 57
64:  br i1 %_8.i, label %bb6.i, label %_RNvMNtNtCscGzNLIcHm04_4core4char7methodsc20is_ascii_punctuation.exit
not:38      !~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  error: no match expected
65:
66: bb6.i: ; preds = %start
67:  %_9.i = icmp samesign ult i32 %0, 65
68:  %_11.i = icmp samesign ugt i32 %0, 90
69:  br i1 %_11.i, label %bb9.i, label %_RNvMNtNtCscGzNLIcHm04_4core4char7methodsc20is_ascii_punctuation.exit
70:
71: bb9.i: ; preds = %bb6.i
72:  %_14.i = icmp samesign ugt i32 %0, 122
73:  %spec.select.v.i = select i1 %_14.i, i32 127, i32 97
74:  %spec.select.i = icmp samesign ult i32 %0, %spec.select.v.i
75:  br label %_RNvMNtNtCscGzNLIcHm04_4core4char7methodsc20is_ascii_punctuation.exit
76:
77: _RNvMNtNtCscGzNLIcHm04_4core4char7methodsc20is_ascii_punctuation.exit: ; preds = %start, %bb6.i, %bb9.i
78:  %_13.sroa.0.0.i = phi i1 [ %_4.sroa.0.0.i, %start ], [ %spec.select.i, %bb9.i ], [ %_9.i, %bb6.i ]
79:  ret i1 %_13.sroa.0.0.i
80: }
81:
82: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable
83: define noundef zeroext i1 @is_ascii_punctuation_u8(i8 noundef %0) unnamed_addr #0 {
84: start:
85:  %1 = add nsw i8 %0, -33
86:  %_4.sroa.0.0.i = icmp ult i8 %1, 15
87:  %_8.i = icmp ugt i8 %0, 57
88:  br i1 %_8.i, label %bb6.i, label %_RNvMs4_NtCscGzNLIcHm04_4core3numh20is_ascii_punctuation.exit
not:45      !~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  error: no match expected
89:
90: bb6.i: ; preds = %start
91:  %_9.i = icmp samesign ult i8 %0, 65
92:  %_11.i = icmp ugt i8 %0, 90
93:  br i1 %_11.i, label %bb9.i, label %_RNvMs4_NtCscGzNLIcHm04_4core3numh20is_ascii_punctuation.exit
94:
95: bb9.i: ; preds = %bb6.i
96:  %_14.i = icmp ugt i8 %0, 122
97:  %spec.select.v.i = select i1 %_14.i, i8 127, i8 97
98:  %spec.select.i = icmp ult i8 %0, %spec.select.v.i
99:  br label %_RNvMs4_NtCscGzNLIcHm04_4core3numh20is_ascii_punctuation.exit
100:
101: _RNvMs4_NtCscGzNLIcHm04_4core3numh20is_ascii_punctuation.exit: ; preds = %start, %bb6.i, %bb9.i
102:  %_13.sroa.0.0.i = phi i1 [ %_4.sroa.0.0.i, %start ], [ %spec.select.i, %bb9.i ], [ %_9.i, %bb6.i ]
103:  ret i1 %_13.sroa.0.0.i
104: }
105:
106: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" }
107:
108: !llvm.module.flags = !{!0, !1}
109: !llvm.ident = !{!2}
110:
111: !0 = !{i32 8, !"PIC Level", i32 2}
112: !1 = !{i32 2, !"RtLibUseGOT", i32 1}
113: !2 = !{!"rustc version 1.96.0-dev"}
>>>>>>
------------------------------------------
 
---- [codegen] tests/codegen-llvm/char-ascii-branchless.rs stdout end ----
```

This is a little over my head to diagnose, so I'm filing a tracking bug so we can get the CI back to green by skipping the test until this bug is fixed.

--- a/llvm/lib/Transforms/Utils/SimplifyCFG.cpp
+++ b/llvm/lib/Transforms/Utils/SimplifyCFG.cpp
@@ -3832,28 +3832,24 @@ static bool foldTwoEntryPHINode(PHINode *PN, const TargetTransformInfo &TTI,
   if (!PN)
     return true;
 
-  // Return true if at least one of these is a 'not', and another is either
-  // a 'not' too, or a constant.
-  auto CanHoistNotFromBothValues = [](Value *V0, Value *V1) {
-    if (!match(V0, m_Not(m_Value())))
-      std::swap(V0, V1);
-    auto Invertible = m_CombineOr(m_Not(m_Value()), m_AnyIntegralConstant());
-    return match(V0, m_Not(m_Value())) && match(V1, Invertible);
-  };
-
   // Don't fold i1 branches on PHIs which contain binary operators or
-  // (possibly inverted) select form of or/ands,  unless one of
-  // the incoming values is an 'not' and another one is freely invertible.
-  // These can often be turned into switches and other things.
-  auto IsBinOpOrAnd = [](Value *V) {
-    return match(
-        V, m_CombineOr(m_BinOp(), m_c_Select(m_ImmConstant(), m_Value())));
+  // (possibly inverted) select form of or/ands if their parameters are
+  // an equality test.
+  auto IsBinOpOrAndEq = [](Value *V) {
+    CmpPredicate Pred;
+    if (match(V, m_CombineOr(
+                     m_CombineOr(
+                         m_BinOp(m_Cmp(Pred, m_Value(), m_Value()), m_Value()),
+                         m_BinOp(m_Value(), m_Cmp(Pred, m_Value(), m_Value()))),
+                     m_c_Select(m_ImmConstant(),
+                                m_Cmp(Pred, m_Value(), m_Value()))))) {
+      return CmpInst::isEquality(Pred);
+    }
+    return false;
   };
   if (PN->getType()->isIntegerTy(1) &&
-      (IsBinOpOrAnd(PN->getIncomingValue(0)) ||
-       IsBinOpOrAnd(PN->getIncomingValue(1)) || IsBinOpOrAnd(IfCond)) &&
-      !CanHoistNotFromBothValues(PN->getIncomingValue(0),
-                                 PN->getIncomingValue(1)))
+      (IsBinOpOrAndEq(PN->getIncomingValue(0)) ||
+       IsBinOpOrAndEq(PN->getIncomingValue(1)) || IsBinOpOrAndEq(IfCond)))
     return Changed;
 
   // If all PHI nodes are promotable, check to make sure that all instructions

--- a/llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll
+++ b/llvm/test/Transforms/PhaseOrdering/X86/vector-reductions.ll
@@ -279,15 +279,14 @@ define i1 @cmp_lt_gt(double %a, double %b, double %c) {
 ; CHECK-NEXT:    [[TMP5:%.*]] = insertelement <2 x double> poison, double [[MUL]], i64 0
 ; CHECK-NEXT:    [[TMP6:%.*]] = shufflevector <2 x double> [[TMP5]], <2 x double> poison, <2 x i32> zeroinitializer
 ; CHECK-NEXT:    [[TMP7:%.*]] = fdiv <2 x double> [[TMP3]], [[TMP6]]
-; CHECK-NEXT:    [[TMP8:%.*]] = fcmp olt <2 x double> [[TMP7]], splat (double 0x3EB0C6F7A0B5ED8D)
+; CHECK-NEXT:    [[TMP8:%.*]] = fcmp uge <2 x double> [[TMP7]], splat (double 0x3EB0C6F7A0B5ED8D)
 ; CHECK-NEXT:    [[SHIFT:%.*]] = shufflevector <2 x i1> [[TMP8]], <2 x i1> poison, <2 x i32> <i32 1, i32 poison>
-; CHECK-NEXT:    [[TMP9:%.*]] = and <2 x i1> [[TMP8]], [[SHIFT]]
-; CHECK-NEXT:    [[OR_COND:%.*]] = extractelement <2 x i1> [[TMP9]], i64 0
+; CHECK-NEXT:    [[OR_COND:%.*]] = or <2 x i1> [[TMP8]], [[SHIFT]]
 ; CHECK-NEXT:    [[TMP10:%.*]] = fcmp ule <2 x double> [[TMP7]], splat (double 1.000000e+00)
 ; CHECK-NEXT:    [[SHIFT2:%.*]] = shufflevector <2 x i1> [[TMP10]], <2 x i1> poison, <2 x i32> <i32 1, i32 poison>
 ; CHECK-NEXT:    [[TMP11:%.*]] = or <2 x i1> [[TMP10]], [[SHIFT2]]
-; CHECK-NEXT:    [[OR_COND1_NOT:%.*]] = extractelement <2 x i1> [[TMP11]], i64 0
-; CHECK-NEXT:    [[RETVAL_0:%.*]] = select i1 [[OR_COND]], i1 false, i1 [[OR_COND1_NOT]]
+; CHECK-NEXT:    [[OR_COND1_NOT:%.*]] = and <2 x i1> [[OR_COND]], [[TMP11]]
+; CHECK-NEXT:    [[RETVAL_0:%.*]] = extractelement <2 x i1> [[OR_COND1_NOT]], i64 0
 ; CHECK-NEXT:    ret i1 [[RETVAL_0]]
 ;
 entry:

--- a/llvm/test/Transforms/SimplifyCFG/extract-cost.ll
+++ b/llvm/test/Transforms/SimplifyCFG/extract-cost.ll
@@ -37,12 +37,8 @@ define i1 @PR32078(<4 x i32> %a, <4 x i32> %b) {
 ; CHECK-NEXT:    [[CMP0_NOT:%.*]] = xor i1 [[CMP0]], true
 ; CHECK-NEXT:    [[CMP1_NOT:%.*]] = xor i1 [[CMP1]], true
 ; CHECK-NEXT:    [[BRMERGE:%.*]] = select i1 [[CMP0_NOT]], i1 true, i1 [[CMP1_NOT]]
-; CHECK-NEXT:    br i1 [[BRMERGE]], label [[EXIT:%.*]], label [[CMP1_TRUE:%.*]]
-; CHECK:       cmp1_true:
 ; CHECK-NEXT:    [[SPEC_SELECT:%.*]] = select i1 [[CMP2]], i1 [[CMP3]], i1 false
-; CHECK-NEXT:    br label [[EXIT]]
-; CHECK:       exit:
-; CHECK-NEXT:    [[R:%.*]] = phi i1 [ false, [[ENTRY:%.*]] ], [ [[SPEC_SELECT]], [[CMP1_TRUE]] ]
+; CHECK-NEXT:    [[R:%.*]] = select i1 [[BRMERGE]], i1 false, i1 [[SPEC_SELECT]]
 ; CHECK-NEXT:    ret i1 [[R]]
 ;
 entry:

--- a/llvm/test/Transforms/SimplifyCFG/switch-transformations-no-lut.ll
+++ b/llvm/test/Transforms/SimplifyCFG/switch-transformations-no-lut.ll
@@ -427,13 +427,9 @@ define i1 @single_value_with_mask(i32 %x) {
 ; TTINOLUT-NEXT:    [[SWITCH_SHIFTED:%.*]] = lshr i64 4294967333, [[SWITCH_MASKINDEX]]
 ; TTINOLUT-NEXT:    [[SWITCH_LOBIT:%.*]] = trunc i64 [[SWITCH_SHIFTED]] to i1
 ; TTINOLUT-NEXT:    [[OR_COND:%.*]] = select i1 [[TMP0]], i1 [[SWITCH_LOBIT]], i1 false
-; TTINOLUT-NEXT:    br i1 [[OR_COND]], label %[[END:.*]], label %[[DEFAULT:.*]]
-; TTINOLUT:       [[DEFAULT]]:
 ; TTINOLUT-NEXT:    [[CMP:%.*]] = icmp eq i32 [[X]], 80
 ; TTINOLUT-NEXT:    [[SEL:%.*]] = select i1 [[CMP]], i1 false, i1 true
-; TTINOLUT-NEXT:    br label %[[END]]
-; TTINOLUT:       [[END]]:
-; TTINOLUT-NEXT:    [[RES:%.*]] = phi i1 [ [[SEL]], %[[DEFAULT]] ], [ false, %[[ENTRY]] ]
+; TTINOLUT-NEXT:    [[RES:%.*]] = select i1 [[OR_COND]], i1 false, i1 [[SEL]]
 ; TTINOLUT-NEXT:    ret i1 [[RES]]
 ;
 entry:

